README.md

The Kigkonsult\Http\Message package provide [PSR] server-side
    
> _Message_ class

> _Stream_ class

> _Response_ class


###### Message class
* implements [Psr\Http\Message\MessageInterface]
* constructor args:  body, headers, protocolVersion

###### Stream class
* implements [Psr\Http\Message\StreamInterface]
* constructor args: content, streamWrapper/resource, mode
* with a (static) factory methods
  *  `factoryFromString( content [, streamWrapper [, mode ]] )`
  *  `factoryFromResource( resource )`

###### Response class
* extends Message
* implements
  *  [Psr\Http\Message\ResponseInterface]
  *  [Fig\Http\Message\StatusCodeInterface]
* constructor args:  body, status, headers

The Response class has an added property: 
* `rawBody`
  *  response body (type mixed) before serialization/encoding

in parallel with 
* `(Psr\Http\Message\)ServerRequest::parsedBody`
* `(Psr\Http\Message\)ServerRequest::body`
 
with corresponding methods:
* `getRawBody()`
  *  Return rawBody
* `isRawBodyEmpty()`
  *  Return bool true on empty rawBody
* `withRawBody()`
  *  Return new instance with rawBody

and two rawBody/body methods
* `getResponseBody()` 
  *  Return rawBody if not empty else body
* `isBodyEmpty()`
  *  Return bool true if (serialized/encoded) body is empty
 
Two statusCode methods are added:

* `assertStatusCode()` 
  *  (static) validates statusCode, throws InvalidArgumentException on error
* `isBodyLessResponse()` 
  *  Return bool true on statusCode 1xx, 204 or 304

 
###### Installation
 
[Composer], from the Command Line:
 
```
composer require kigkonsult/http-master:dev-master
```
 
Composer, in your `composer.json`:
 
``` json
{
    "require": {
        "kigkonsult/http-master": "dev-master"
    }
}
```
 
Composer, acquire access
``` php
<?php
use Kigkonsult\Http\Message\Stream;
use Kigkonsult\Http\Message\Message;
use Kigkonsult\Http\Message\Response;
...
include 'vendor/autoload.php';
```
 
 
Otherwise , download and acquire..
 
``` php
<?php
use Kigkonsult\Http\Message\Stream;
use Kigkonsult\Http\Message\Message;
use Kigkonsult\Http\Message\Response;
...
include 'pathToSource/http-master/autoload.php';
```

###### Support

For support use [github.com] http-master. Non-emergence support issues are, unless sponsored, fixed in due time.


###### Sponsorship

Donation using <a href="https://paypal.me/kigkonsult" rel="nofollow">paypal.me/kigkonsult</a> are appreciated.
For invoice, <a href="mailto:ical@kigkonsult.se">please e-mail</a>.

###### License

This project is licensed under the LGPLv3 License


[Composer]:https://getcomposer.org/
[github.com]:https://github.com/iCalcreator/PcGen
[Psr]:https://github.com/php-fig/http-message
[Psr\Http\Message\MessageInterface]:https://github.com/php-fig/http-message
[Psr\Http\Message\StreamInterface]:https://github.com/php-fig/http-message
[Psr\Http\Message\ResponseInterface]:https://github.com/php-fig/http-message
[Fig\Http\Message\StatusCodeInterface]:https://github.com/php-fig/http-message-util


ReleaseNotes

1.1   PHP 7 (PHPStan/strict types)




test/StreamTest.php
<?php
/**
 * http-message, a Psr\Http\Message implementation
 *
 * This file is part of http-message.
 *
 * @author    Kjell-Inge Gustafsson, kigkonsult <ical@kigkonsult.se>
 * @copyright 2019-2021 Kjell-Inge Gustafsson, kigkonsult, All rights reserved
 * @link      https://kigkonsult.se
 * @license   Subject matter of licence is the software http-message.
 *            The above copyright, link and this licence notice shall be
 *            included in all copies or substantial portions of the http-message.
 *
 *            http-message is free software: you can redistribute it and/or modify
 *            it under the terms of the GNU Lesser General Public License as
 *            published by the Free Software Foundation, either version 3 of
 *            the License, or (at your option) any later version.
 *
 *            http-message is distributed in the hope that it will be useful,
 *            but WITHOUT ANY WARRANTY; without even the implied warranty of
 *            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *            GNU Lesser General Public License for more details.
 *
 *            You should have received a copy of the GNU Lesser General Public License
 *            along with http-message. If not, see <https://www.gnu.org/licenses/>.
 */
declare( strict_types = 1 );
namespace Kigkonsult\Http\Message;

use PHPUnit\Framework\TestCase;
use Exception;

/**
 * class StreamTest
 *
 * @author      Kjell-Inge Gustafsson <ical@kigkonsult.se>
 */
class StreamTest extends TestCase
{
    private static $loremIpsum = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer at ligula quis odio pellentesque aliquam eget a lorem. Nulla et justo in quam ullamcorper faucibus. Fusce iaculis sagittis ligula quis scelerisque. Proin vitae enim cursus, ullamcorper sapien non, consectetur justo. Maecenas ligula velit, semper vitae posuere nec, imperdiet at metus. Nulla egestas, turpis lacinia aliquet iaculis, tellus velit imperdiet leo, dignissim commodo quam lacus eget libero. Integer nec sodales sapien. Nunc ullamcorper malesuada sem in condimentum. Etiam facilisis, nisl non eleifend dapibus, ipsum nisi efficitur sapien, eu placerat nisl risus id sapien. Duis vulputate ut tellus sagittis feugiat. Aenean tincidunt sodales est, ac condimentum augue aliquet finibus. Maecenas ultricies, tortor non volutpat imperdiet, sapien turpis sodales sapien, sed mollis diam elit eu arcu. Quisque placerat egestas tempus. Morbi libero sapien, molestie quis ligula nec, eleifend mattis turpis.In lacinia, tortor ut vehicula viverra, arcu nisl ornare metus, at porta quam est eu dolor. Donec pretium mauris nibh, vel egestas lorem bibendum non. Cras tincidunt consequat nisi. Maecenas volutpat turpis non urna laoreet, sed molestie metus hendrerit. Nam a ultricies justo, ut ullamcorper ipsum. Praesent sed consequat quam. Suspendisse id lacinia quam.
Phasellus in neque sed nisl hendrerit facilisis. Nulla eu elit gravida, condimentum nulla id, fringilla enim. Nulla maximus dui id vehicula sagittis. Vestibulum tincidunt porta risus, eu mattis lectus laoreet eget. Sed tristique sit amet dui venenatis condimentum. Donec lobortis diam sed lectus blandit posuere. Quisque odio arcu, condimentum nec tempor ut, faucibus sit amet sem. Nullam ut porta arcu, accumsan tempus orci. Mauris id eros sit amet urna fringilla pulvinar. Aenean vitae ligula id ante ultrices consequat.
Vestibulum eget cursus enim. Curabitur nec elementum ex. Vestibulum non tortor purus. Quisque a mi sit amet erat consequat ultrices at in lectus. Proin urna sapien, ornare nec tempor non, dignissim bibendum ante. Etiam sem dolor, dapibus eget interdum vitae, dapibus a tellus. Duis volutpat at erat quis volutpat. Vestibulum id sollicitudin magna, quis mattis turpis. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Nullam pellentesque mollis aliquet. Ut tincidunt velit dui, quis accumsan dui auctor et. Vivamus eu arcu eu risus fermentum pharetra. In luctus dui nulla, at finibus nibh ornare at. Sed porttitor consequat accumsan. Sed vel diam dolor. Cras vitae malesuada sapien.
Quisque vel odio id tortor congue malesuada. Praesent a ex turpis. Donec pharetra elit ut bibendum semper. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Mauris feugiat elit nec sapien imperdiet, in condimentum nibh sodales. Sed nec accumsan felis. Nulla leo turpis, egestas sed tincidunt eu, bibendum in libero.';

    private static $fileName = [];

    public static function tearDownAfterClass()
    {
        foreach( self::$fileName as $file ) {
            unlink( $file );
        }
        self::$fileName = null;
    }

    /**
     * @param string $data
     * @return resource
     */
    private function getResourceHandle( string $data )
    {
        $fileName         = tempnam( sys_get_temp_dir(), "test" );
        self::$fileName[] = $fileName;
        $handle           = fopen( $fileName, "wb+" );
        fwrite( $handle, $data );
        return $handle;
    }

    /**
     * streamTest1+2 provider
     */
    public function streamTestProvider() : array
    {

        $dataArr       = [];

        $dataArr[] = [
            11,
            null,
            null,
            null,
            null,
        ];

        $dataArr[] = [
            12,
            self::$loremIpsum,
            null,
            null,
            self::$loremIpsum,
        ];

        $dataArr[] = [
            21,
            null,
            'php://memory',
            'wb+',
            null,
        ];

        $dataArr[] = [
            22,
            null,
            'php://memory',
            null,
            null,
        ];

        $dataArr[] = [
            23,
            self::$loremIpsum,
            'php://memory',
            'wb+',
            self::$loremIpsum,
        ];

        $dataArr[] = [
            24,
            self::$loremIpsum,
            'php://memory',
            null,
            self::$loremIpsum,
        ];

        $dataArr[] = [
            31,
            null,
            $this->getResourceHandle( '' ),
            null,
            '',
        ];

        $dataArr[] = [
            32,
            null,
            $this->getResourceHandle( self::$loremIpsum ),
            null,
            self::$loremIpsum,
        ];

        return $dataArr;
    }

    /**
     * test Stream instance
     *
     * @test
     * @dataProvider streamTestProvider
     *
     * @param int    $case
     * @param string $content
     * @param string $streamWrapper
     * @param string $mode
     * @param string $expected
     */
    public function streamTest1(
        int $case,
        $content,
        $streamWrapper,
        $mode,
        $expected
    )
    {
        $this->streamTestX(
            100 + $case,
            new Stream( $content, $streamWrapper, $mode ),
            $expected
        );
    }

    /**
     * test Stream factory
     *
     * @test
     * @dataProvider streamTestProvider
     *
     * @param int    $case
     * @param string $content
     * @param string $streamWrapper
     * @param string $mode
     * @param string $expected
     */
    public function streamTest2(
        int $case,
        $content,
        $streamWrapper,
        $mode,
        $expected
    )
    {
        $this->streamTestX(
            200 + $case,
            Stream::factory( $content, $streamWrapper, $mode ),
            $expected
        );
    }

    /**
     * test Stream factoryFromString / factoryFromResource
     *
     * @test
     * @dataProvider streamTestProvider
     *
     * @param int    $case
     * @param string $content
     * @param string $streamWrapper
     * @param string $mode
     * @param string $expected
     */
    public function streamTest345(
        int $case,
        $content,
        $streamWrapper,
        $mode,
        $expected
    )
    {
        switch( true ) {
            case is_resource( $streamWrapper ) :
                $this->streamTestX(
                    300 + $case,
                    Stream::factoryFromResource( $streamWrapper ),
                    $expected
                );
                break;
            case ( ! empty( $content )) :
                $this->streamTestX(
                    400 + $case,
                    Stream::factoryFromString( $content  ),
                    $expected
                );
                break;
            default :
                $this->streamTestX(
                    500 + $case,
                    Stream::factory( $content, $streamWrapper, $mode ),
                    $expected
                );
                break;
        } // end switch
    }

    /**
     * @param int    $case
     * @param stream $stream
     * @param string $expected
     */
    public function streamTestX(
        int $case,
        $stream,
        $expected
    )
    {
        static $FMTERR = 'Error %d in case #%d';

        $this->assertTrue(
            $stream->isReadable(),
            sprintf( $FMTERR, 1, $case )
        );
        $this->assertTrue(
            $stream->isWritable(),
            sprintf( $FMTERR, 2, $case )
        );

        $this->assertTrue(
            $stream->isSeekable(),
            sprintf( $FMTERR, 3, $case )
        );

        $stream->rewind();
        $this->assertEmpty(
            $stream->tell(),
            sprintf( $FMTERR, 4, $case )
        );

        $this->assertEquals(
            empty( $expected ),
            Stream::isStreamEmpty( $stream ),
            sprintf( $FMTERR, 5, $case )
        );

        $contentLength = strlen( $expected ?? '' );
        $this->assertEquals(
            $contentLength,
            $stream->getSize(),
            sprintf( $FMTERR, 6, $case )
        );

        if( ! empty( $contentLength )) {
            $stream->seek( 0 );
            $this->assertEquals(
                $expected,
                $stream->read( $contentLength ),
                sprintf( $FMTERR, 7, $case )
            );
        }

        $stream->rewind();
        $this->assertEquals(
            $expected,
            $stream->getContents(),
            sprintf( $FMTERR, 7, $case )
        );
        $this->assertTrue(
            $stream->eof(),
            sprintf( $FMTERR, 8, $case )
        );

        $stream->detach();
    }

    /**
     * streamTest4 provider
     */
    public function streamTest4Provider() : array
    {

        $dataArr       = [];

        $dataArr[] = [
            701,
            'content',
            'php://memory',
            'r'
        ];

        return $dataArr;
    }

    /**
     * test Stream resource exceptions
     *
     * @test
     * @dataProvider streamTest4Provider
     *
     * @param int    $case
     * @param string $content
     * @param string wrapper
     * @param string $mode
     */
    public function streamTest7(
        $case,
        $content,
        $wrapper,
        $mode
    ) {
        static $FMTERR = 'Error in case #%d';
        $error = false;
        try {
            $stream = new Stream( $content, $wrapper, $mode );
        }
        catch( Exception $e ) {
            $error = true;
        }
        $this->assertTrue(
            $error,
            sprintf( $FMTERR, $case )
        );
    }
}


test/ResponseTest.php
<?php
/**
 * http-message, a Psr\Http\Message implementation
 *
 * This file is part of http-message.
 *
 * @author    Kjell-Inge Gustafsson, kigkonsult <ical@kigkonsult.se>
 * @copyright 2019-2021 Kjell-Inge Gustafsson, kigkonsult, All rights reserved
 * @link      https://kigkonsult.se
 * @license   Subject matter of licence is the software http-message.
 *            The above copyright, link and this licence notice shall be
 *            included in all copies or substantial portions of the http-message.
 *
 *            http-message is free software: you can redistribute it and/or modify
 *            it under the terms of the GNU Lesser General Public License as
 *            published by the Free Software Foundation, either version 3 of
 *            the License, or (at your option) any later version.
 *
 *            http-message is distributed in the hope that it will be useful,
 *            but WITHOUT ANY WARRANTY; without even the implied warranty of
 *            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *            GNU Lesser General Public License for more details.
 *
 *            You should have received a copy of the GNU Lesser General Public License
 *            along with http-message. If not, see <https://www.gnu.org/licenses/>.
 */
declare( strict_types = 1 );
namespace Kigkonsult\Http\Message;

use PHPUnit\Framework\TestCase;
use InvalidArgumentException;
use Exception;

/**
 * class ResponseTest
 *
 * @author      Kjell-Inge Gustafsson <ical@kigkonsult.se>
 */
class ResponseTest extends TestCase
{

    /**
     * @var string
     */
    private $fileName = null;

    protected function tearDown()
    {
        if( ! empty( $this->fileName ) && is_file( $this->fileName )) {
            unlink( $this->fileName );
        }
    }

    /**
     * @param null|mixed $data
     * @return resource
     */
    private function getResourceHandle( $data )
    {
        $this->fileName = tempnam( sys_get_temp_dir(), "test" );
        $handle         = fopen( $this->fileName, "wb+" );
        fwrite( $handle, (string) $data ?? '' );
        return $handle;
    }

    /**
     * Testing response status
     *
     * @test
     * @ // expectedException InvalidArgumentException
     */
    public function testSetStatusCode()
    {
        $data     = '1!2"3#4¤5%6&7/8(9)0=+?*.:,;';
        $this->getResourceHandle( $data );
        try {
            $stream = new Stream( null, fopen( $this->fileName, 'wb+' ));
            $reponse = new Response( $stream, 1234, [] );
            $this->expectException( InvalidArgumentException::class );
        }
        catch( InvalidArgumentException $e ) {}
        catch( Exception $e ) {}
        $this->assertTrue( true );
    }

    /**
     * test empty rawBody
     *
     * @test
     */
    public function testrawBody1()
    {
        $response = new Response();
        $response = $response->withRawBody();
        $this->assertEmpty( $response->getRawBody());
        $this->assertTrue( $response->isRawBodyEmpty());
        $this->assertEquals( null, $response->getResponseBody());
    }

    /**
     * testBody2 provider
     */
    public function body2Provider() : array
    {

        $dataArr       = [];

        $dataArr[] = [
            11, null, null, null    // expects rawBody
        ];

        $dataArr[] = [
            21, 1, null, 1          // expects rawBody
        ];

        $dataArr[] = [
            22, null, 1, 1          // expects body
        ];

        $dataArr[] = [
            23, 1, 2, 1             // expects rawBody
        ];

        $dataArr[] = [
            31, 0, null, 0          // expects rawBody
        ];

        $dataArr[] = [
            32, null, 0, 0          // expects body
        ];

        $dataArr[] = [
            33, 0, 1, 0             // expects rawBody
        ];

        $dataArr[] = [
            41, false, null, false  // expects rawBody
        ];

        $dataArr[] = [
            42, null, false, false  // expects body
        ];

        $dataArr[] = [
            43, false, 1, false     // expects rawBody
        ];

        $dataArr[] = [
            61, true, null, true    // expects rawBody
        ];

        $dataArr[] = [
            62, null, true, true    // expects body
        ];

        $dataArr[] = [
            63, true, 0, true       // expects rawBody
        ];

        $dataArr[] = [
            71, 1, '', 1            // expects rawBody
        ];

        $dataArr[] = [
            72, '', 1, 1            // expects body
        ];

        return $dataArr;
    }

    /**
     * test empty body
     *
     * @test
     * @dataProvider body2Provider
     *
     * @param int    $case
     * @param string $rawBody
     * @param string $body
     * @param string $expected
     */
    public function testBody2(
        int $case,
        $rawBody,
        $body,
        $expected
    )
    {
        $response = new Response();
        $response = $response->withRawBody( $rawBody );

        $response = $response->withBody(
            Stream::factoryFromResource( $this->getResourceHandle( $body ))
        );

        $this->assertEquals(
            $expected,
            $response->getResponseBody(),
            'error in case #' . $case .
            ' rawBody: '  . var_export( $rawBody,  true ) .
            ' body: '     . var_export( $body,     true ) .
            ' expected: ' . var_export( $expected, true )
        );

        if( empty( $rawBody ) && ( '0' != $rawBody )) {
            $this->assertTrue( $response->isRawBodyEmpty());
        }

        if( empty( $body ) && ( '0' != $body )) {
            $this->assertTrue( $response->isBodyEmpty());
        }
    }

    /**
     * test response
     *
     * @test
     */
    public function testrawBody3()
    {
        $data     = '1!2"3#4¤5%6&7/8(9)0=+?*.:,;';
        $response = new Response();
        $response = $response->withRawBody( $data )
                             ->withStatus( Response::STATUS_INTERNAL_SERVER_ERROR );

        $this->assertEquals( Response::STATUS_INTERNAL_SERVER_ERROR, $response->getStatusCode());
        $this->assertEquals( 'Internal Server Error', $response->getReasonPhrase());

        $this->assertEquals( $data, $response->getRawBody());
        $this->assertEquals( $data, $response->getResponseBody());

    }

    /**
     * testisBodyLessResponse provider
     */
    public function isBodyLessResponseProvider() : array
    {
        $dataArr = [];

        $dataArr[] = [
                new Response(),
                false,
        ];

        $dataArr[] = [
                new Response( 'php://memory', Response::STATUS_CONTINUE ),
                true,
        ];

        $dataArr[] = [
                new Response( 'php://memory', Response::STATUS_NO_CONTENT ),
                true,
        ];

        $dataArr[] = [
                new Response( 'php://memory', Response::STATUS_NOT_MODIFIED ),
                true,
        ];

        $dataArr[] = [
                new Response( 'php://memory', Response::STATUS_INTERNAL_SERVER_ERROR ),
                false,
        ];

        return $dataArr;
    }

    /**
     * test Response statusCodes
     *
     * @test
     * @dataProvider isBodyLessResponseProvider
     * @param ResponseInterface $response
     * @param bool              $expected
     */
    public function testisBodyLessResponse(
        ResponseInterface $response,
        bool $expected
    )
    {
        $this->assertEquals( $expected, $response->isBodyLessResponse());
    }
}


src/Response.php
<?php
/**
 * http-message, a Psr\Http\Message implementation
 *
 * This file is part of http-message.
 *
 * @author    Kjell-Inge Gustafsson, kigkonsult <ical@kigkonsult.se>
 * @copyright 2019-2021 Kjell-Inge Gustafsson, kigkonsult, All rights reserved
 * @link      https://kigkonsult.se
 * @license   Subject matter of licence is the software http-message.
 *            The above copyright, link and this licence notice shall be
 *            included in all copies or substantial portions of the http-message.
 *
 *            http-message is free software: you can redistribute it and/or modify
 *            it under the terms of the GNU Lesser General Public License as
 *            published by the Free Software Foundation, either version 3 of
 *            the License, or (at your option) any later version.
 *
 *            http-message is distributed in the hope that it will be useful,
 *            but WITHOUT ANY WARRANTY; without even the implied warranty of
 *            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *            GNU Lesser General Public License for more details.
 *
 *            You should have received a copy of the GNU Lesser General Public License
 *            along with http-message. If not, see <https://www.gnu.org/licenses/>.
 */
declare( strict_types = 1 );
namespace Kigkonsult\Http\Message;

use Psr\Http\Message\StreamInterface;
use InvalidArgumentException;

use function gettype;
use function is_scalar;
use function sprintf;

/**
 * Implement Psr\Http\Message\ResponseInterface with
 *   new property: rawBody (mixed)
 *   methods
 *     getRawBody()
 *     isRawBodyEmpty()
 *     withRawBody()
 *     isBodyEmpty()
 *
 *     assertStatusCode()
 *     isStatusCodeValid()
 *     getResponseBody()
 *     isBodyLessResponse()
 */
/**
 * HTTP response encapsulation.
 *
 * Responses are considered immutable; all methods that might change state are
 * implemented such that they retain the internal state of the current
 * message and return a new instance that contains the changed state.
 *
 *
 */
class Response extends Message implements ResponseInterface
{
    /**
     * @const string
     */
    const MIN_STATUS_CODE_VALUE = 100;
    const MAX_STATUS_CODE_VALUE = 599;

    /**
     * Map of standard HTTP status code/reason phrases
     *
     * @var array
     */
    private $phrases = [
        // INFORMATIONAL CODES
        self::STATUS_CONTINUE            => 'Continue',                  // 100;
        self::STATUS_SWITCHING_PROTOCOLS => 'Switching Protocols',       // 101
        self::STATUS_PROCESSING          => 'Processing',                // 102
        self::STATUS_EARLY_HINTS         => 'Early Hints',               // 203
        // SUCCESS CODES
        self::STATUS_OK                  => 'OK',                        // 200
        self::STATUS_CREATED             => 'Created',                   // 201
        self::STATUS_ACCEPTED            => 'Accepted',                  // 202
        self::STATUS_NON_AUTHORITATIVE_INFORMATION                       // 203
                                         => 'Non-Authoritative Information',
        self::STATUS_NO_CONTENT          => 'No Content',                // 204
        self::STATUS_RESET_CONTENT       => 'Reset Content',             // 205
        self::STATUS_PARTIAL_CONTENT     => 'Partial Content',           // 206
        self::STATUS_MULTI_STATUS        => 'Multi-Status',              // 207
        self::STATUS_ALREADY_REPORTED    => 'Already Reported',          // 208
        self::STATUS_IM_USED             => 'IM Used',                   // 226
        // REDIRECTION CODES
        self::STATUS_MULTIPLE_CHOICES    => 'Multiple Choices',          // 300
        self::STATUS_MOVED_PERMANENTLY   => 'Moved Permanently',         // 301
        self::STATUS_FOUND               => 'Found',                     // 302
        self::STATUS_SEE_OTHER           => 'See Other',                 // 303
        self::STATUS_NOT_MODIFIED        => 'Not Modified',              // 304
        self::STATUS_USE_PROXY           => 'Use Proxy',                 // 305
        self::STATUS_RESERVED            => 'Switch Proxy',              // 306, Deprecated to '(Unused)'
        self::STATUS_TEMPORARY_REDIRECT  => 'Temporary Redirect',        // 307
        self::STATUS_PERMANENT_REDIRECT  => 'Permanent Redirect',        // 308
        // CLIENT ERROR
        self::STATUS_BAD_REQUEST         => 'Bad Request',               // 400
        self::STATUS_UNAUTHORIZED        => 'Unauthorized',              // 401
        self::STATUS_PAYMENT_REQUIRED    => 'Payment Required',          // 402
        self::STATUS_FORBIDDEN           => 'Forbidden',                 // 403
        self::STATUS_NOT_FOUND           => 'Not Found',                 // 404
        self::STATUS_METHOD_NOT_ALLOWED  => 'Method Not Allowed',        // 405
        self::STATUS_NOT_ACCEPTABLE      => 'Not Acceptable',            // 406
        self::STATUS_PROXY_AUTHENTICATION_REQUIRED                       // 407
                                         => 'Proxy Authentication Required',
        self::STATUS_REQUEST_TIMEOUT     => 'Request Timeout',           // 408
        self::STATUS_CONFLICT            => 'Conflict',                  // 409
        self::STATUS_GONE                => 'Gone',                      // 410
        self::STATUS_LENGTH_REQUIRED     => 'Length Required',           // 411
        self::STATUS_PRECONDITION_FAILED => 'Precondition Failed',       // 412
        self::STATUS_PAYLOAD_TOO_LARGE   => 'Payload Too Large',         // 413
        self::STATUS_URI_TOO_LONG        => 'URI Too Long',              // 414
        self::STATUS_UNSUPPORTED_MEDIA_TYPE                              // 415
                                         => 'Unsupported Media Type',
        self::STATUS_RANGE_NOT_SATISFIABLE                               // 416
                                         => 'Range Not Satisfiable',
        self::STATUS_EXPECTATION_FAILED  => 'Expectation Failed',        // 417
        self::STATUS_IM_A_TEAPOT         => 'I\'m a teapot',             // 418
        self::STATUS_MISDIRECTED_REQUEST => 'Misdirected Request',       // 421
        self::STATUS_UNPROCESSABLE_ENTITY => 'Unprocessable Entity',     // 422
        self::STATUS_LOCKED              => 'Locked',                    // 423
        self::STATUS_FAILED_DEPENDENCY   => 'Failed Dependency',         // 424
        self::STATUS_TOO_EARLY           => 'Too early',                 // 425
        self::STATUS_UPGRADE_REQUIRED    => 'Upgrade Required',          // 426
        self::STATUS_PRECONDITION_REQUIRED
                                         => 'Precondition Required',     // 428
        self::STATUS_TOO_MANY_REQUESTS   => 'Too Many Requests',         // 429
        self::STATUS_REQUEST_HEADER_FIELDS_TOO_LARGE                     // 431
                                         => 'Request Header Fields Too Large',
                                                                         // 444
        444                              => 'Connection Closed Without Response',
        self::STATUS_UNAVAILABLE_FOR_LEGAL_REASONS                       // 451
                                         => 'Unavailable For Legal Reasons',
        // SERVER ERROR
        499                              => 'Client Closed Request',     // 499
        self::STATUS_INTERNAL_SERVER_ERROR => 'Internal Server Error',   // 500
        self::STATUS_NOT_IMPLEMENTED     => 'Not Implemented',           // 501
        self::STATUS_BAD_GATEWAY         => 'Bad Gateway',               // 502
        self::STATUS_SERVICE_UNAVAILABLE => 'Service Unavailable',       // 503
        self::STATUS_GATEWAY_TIMEOUT     => 'Gateway Timeout',           // 504
        self::STATUS_VERSION_NOT_SUPPORTED                               // 505
                                         => 'HTTP Version Not Supported',
        self::STATUS_VARIANT_ALSO_NEGOTIATES                             // 506
                                         => 'Variant Also Negotiates',
        self::STATUS_INSUFFICIENT_STORAGE                                // 507
                                         => 'Insufficient Storage',
        self::STATUS_LOOP_DETECTED       => 'Loop Detected',             // 508
        self::STATUS_NOT_EXTENDED        => 'Not Extended',              // 510
        self::STATUS_NETWORK_AUTHENTICATION_REQUIRED                     // 511
                                         => 'Network Authentication Required',
                                                                         // 599
        599                              => 'Network Connect Timeout Error',
    ];

    /**
     * @var string
     */
    private $reasonPhrase = '';

    /**
     * @var int
     */
    private $statusCode;

    /**
     * @var mixed  the unserialized and uncompressed message body
     */
    private $rawBody;

    /**
     * @param null|string|resource|StreamInterface $body
     * @param null|int                    $status  Status code for the response, if any.
     * @param null|array                  $headers for the response, if any.
     * @throws InvalidArgumentException on any invalid element.
     */
    public function __construct(
        $body    = null,
        $status  = null,
        $headers = []
    )
    {
        parent::__construct( $body, (array) $headers );
        $this->setStatusCode( $status ?? self::STATUS_OK );
    }

    /**
     * {@inheritdoc}
     */
    public function getStatusCode() : int
    {
        return $this->statusCode;
    }

    /**
     * {@inheritdoc}
     */
    public function getReasonPhrase() : string
    {
        if( empty( $this->reasonPhrase ) &&
            isset( $this->phrases[$this->statusCode] )) {
            $this->reasonPhrase = $this->phrases[$this->statusCode];
        }
        return $this->reasonPhrase;
    }

    /**
     * {@inheritdoc}
     */
    public function withStatus( $code, $reasonPhrase = '' ) : ResponseInterface
    {
        $new = clone $this;
        $new->setStatusCode( $code );
        $new->reasonPhrase = $reasonPhrase;
        return $new;
    }

    /**
     * Set a valid status code.
     *
     * @param int $code
     * @throws InvalidArgumentException on an invalid status code.
     * @return static
     */
    private function setStatusCode( int $code ) : self
    {
        self::assertStatusCode( $code );
        $this->statusCode = $code;
        return $this;
    }

    /**
     * Assert status code.
     *
     * @param int $code
     * @throws InvalidArgumentException on an invalid status code.
     */
    public static function assertStatusCode( int $code )
    {
        static $FMT = 'Invalid status code "%s" (%d - %d)';
        if( ! self::isStatusCodeValid( $code )) {
            throw new InvalidArgumentException(
                sprintf(
                    $FMT,
                    ( is_scalar($code) ? $code : gettype( $code )),
                    static::MIN_STATUS_CODE_VALUE,
                    static::MAX_STATUS_CODE_VALUE
                )
            );
        }
    }

    /**
     * Return true i statusCode is int and 1xx-599
     *
     * @param mixed $code
     * @return bool
     */
    public static function isStatusCodeValid( $code ) : bool
    {
        return (( $code == intval( $code ))  &&
            ( $code >= static::MIN_STATUS_CODE_VALUE ) &&
            ( $code <= static::MAX_STATUS_CODE_VALUE ));
    }

    /**
     * {@inheritdoc}
     */
    public function getRawBody()
    {
        return $this->rawBody;
    }

    /**
     * {@inheritdoc}
     */
    public function isRawBodyEmpty() : bool
    {
        // return ( empty( $this->rawBody ));
        return ( empty( $this->rawBody ) && ( '0' != $this->rawBody ));
    }

    /**
     * {@inheritdoc}
     */

    public function withRawBody( $rawBody = null ) : ResponseInterface
    {
        $new          = clone $this;
        $new->rawBody = $rawBody;
        return $new;
    }

    /**
     * {@inheritdoc}
     */
    public function isBodyEmpty() : bool
    {
        return empty( $this->getBody()->getSize());
        /*
        $size = $this->getBody()->getSize();
        echo __FUNCTION__ . ' size=' . $size . PHP_EOL; // test ###
        return empty( $size );
        */
    }

    /**
     * {@inheritdoc}
     */
    public function getResponseBody()
    {
        if( ! $this->isRawBodyEmpty()) {
            return $this->getRawBody();
        }
        if( ! $this->isBodyEmpty()) {
            $stream = $this->getBody();
            $stream->rewind();
            return $stream->getContents();
        }
        return null;
    }

    /**
     * {@inheritdoc}
     */
    public function isBodyLessResponse() : bool
    {
        $statusCode = $this->getStatusCode();
        switch( true ) {
            case (( self::STATUS_CONTINUE <= $statusCode ) && ( self::STATUS_OK > $statusCode )) :
                return true; // 1xx
            case ( self::STATUS_NO_CONTENT == $statusCode ) :
                return true; // 204
            case ( self::STATUS_NOT_MODIFIED == $statusCode ) :
                return true; // 304
            default :
                break;
        }
        return false;
    }
}
